core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java

```diff
@@ -3443,12 +3444,32 @@ private RexNode convertUsing(SqlValidatorNamespace leftNamespace,
           throw new AssertionError("field " + name + " is not found in "
               + rowType + " with " + nameMatcher);
         }
+
+        fieldTypes.add(field.getType());
```

Comment:
Can we just use `rowType` here?

Reply:
This list just contains two fields, one from the left operand of the join and one from the right operand.
The list is created for each pair of fields that are joined on.
It's not really row type.